File: LWHeroOC/Classes/CascadePreprocessor.m
//
//  CascadePreprocessor.m
//  Hero-ObjectiveC
//
//  Created by luca.li on 2017/1/24.
//  Copyright © 2017年 luca. All rights reserved.
//

#import "CascadePreprocessor.h"

@interface CascadePreprocessor ()

@property (nonatomic, assign) CascadeDirection direction;
@property (nonatomic, assign) NSComparator comparator;

@end

@implementation CascadePreprocessor

- (instancetype)initWithDirectionString:(NSString *)string {
    
    if (self = [super init]) {
        
        __weak typeof(self) weakSelf = self;
        self.comparator = ^NSComparisonResult(UIView *view1, UIView *view2) {
            switch (weakSelf.direction) {
                case CascadeDirectionTopToBottom:
                    return CGRectGetMinY(view1.frame) < CGRectGetMinY(view2.frame);
                    break;
                case CascadeDirectionBottomToTop:
                    return (CGRectGetMaxY(view1.frame) == CGRectGetMaxY(view2.frame) ?
                            CGRectGetMaxX(view1.frame) > CGRectGetMaxX(view2.frame) :
                            CGRectGetMaxY(view1.frame) > CGRectGetMaxY(view2.frame));
                    break;
                case CascadeDirectionLeftToRight:
                    return CGRectGetMinX(view1.frame) < CGRectGetMinX(view2.frame);
                    break;
                case CascadeDirectionRightToLeft:
                    return CGRectGetMaxX(view1.frame) < CGRectGetMaxX(view2.frame);
                    break;
                case CascadeDirectionRadial:
                    return ([weakSelf distanceFrom:view1.center to:weakSelf.center] <
                            [weakSelf distanceFrom:view2.center to:weakSelf.center]);
                    break;
                case CascadeDirectionInverseRadial:
                    return ([weakSelf distanceFrom:view2.center to:weakSelf.center] <
                            [weakSelf distanceFrom:view1.center to:weakSelf.center]);
                    break;
                default:
                    break;
            }
        };
        
        if ([string isEqualToString:@"bottomToTop"]) {
            self.direction = CascadeDirectionBottomToTop;
            return self;
        }
        if ([string isEqualToString:@"leftToRight"]) {
            self.direction = CascadeDirectionBottomToTop;
            return self;
        }
        if ([string isEqualToString:@"rightToLeft"]) {
            self.direction = CascadeDirectionBottomToTop;
            return self;
        }
        if ([string isEqualToString:@"topToBottom"]) {
            self.direction = CascadeDirectionBottomToTop;
            return self;
        }
        return nil;
    }
    
    return self;
}

- (void)processFromViews:(NSArray *)fromviews toViews:(NSArray *)toviews {
    [self processViews:fromviews];
    [self processViews:toviews];
}

- (void)processViews:(NSArray <UIView *> *)views {
    [views enumerateObjectsUsingBlock:^(UIView * _Nonnull fv, NSUInteger viewIndex, BOOL * _Nonnull stop) {
        if ([self.context stateOfView:fv].cascade) {
            
            NSArray *cascadeParam = [self.context stateOfView:fv].cascade;
            NSTimeInterval deltaTime = [cascadeParam[0] doubleValue];
            BOOL delayMatchedViews = [cascadeParam[2] boolValue];
            
            UIView *parentView = fv;
            if ([fv isKindOfClass:[UITableView class]]) {
                UIView *wrapperView = [fv.subviews firstObject];
                if (wrapperView) {
                    parentView = wrapperView;
                }
            }
            
            NSArray *sortedSubviews = parentView.subviews;
            NSMutableArray *unpairedViews = [NSMutableArray array];
            [sortedSubviews enumerateObjectsUsingBlock:^(UIView * _Nonnull view, NSUInteger idx, BOOL * _Nonnull stop) {
                if (![self.context pairedViewForView:view]) {
                    [unpairedViews addObject:view];
                }
            }];
            [sortedSubviews sortedArrayUsingComparator:self.comparator];
            
            NSTimeInterval initialDelay = [[self.context stateOfView:fv].delay doubleValue];
            [sortedSubviews enumerateObjectsUsingBlock:^(UIView * _Nonnull v, NSUInteger i, BOOL * _Nonnull stop) {
                NSTimeInterval delay = (NSTimeInterval)i * deltaTime + initialDelay;
                [self.context stateOfView:v].delay = @(delay);
            }];
            
            if (delayMatchedViews) {
                for (NSInteger i = viewIndex + 1; i < views.count; i ++) {
                    UIView *otherView = views[i];
                    if (otherView.superview == fv.superview) {
                        break;
                    }
                    UIView *pairedView = [self.context pairedViewForView:otherView];
                    if (parentView) {
                        NSTimeInterval delay = (NSTimeInterval)sortedSubviews.count * deltaTime + initialDelay;
                        [self.context stateOfView:otherView].delay = @(delay);
                        [self.context stateOfView:pairedView].delay = @(delay);
                    }
                }
            }
        }
    }];
}

#pragma mark - Private
- (CGFloat)distanceFrom:(CGPoint)a to:(CGPoint)b {
    return sqrt(pow(a.x - b.x, 2) + pow(a.y - b.y, 2));
}
@end

File: LWHero_swift/LWHero/Classes/CascadePreprocessor.swift
//
//  CascadePreprocessor.swift
//  LWHero
//
//  Created by Swift Port
//  Copyright © 2025. All rights reserved.
//

import Foundation
import UIKit

// MARK: - CascadeDirection Enum

public enum CascadeDirection {
    case topToBottom
    case bottomToTop
    case leftToRight
    case rightToLeft
    case radial(center: CGPoint)
    case inverseRadial(center: CGPoint)

    init?(string: String, center: CGPoint?) {
        switch string.lowercased() {
        case "toptobottom":
            self = .topToBottom
        case "bottomtotop":
            self = .bottomToTop
        case "lefttoright":
            self = .leftToRight
        case "righttoleft":
            self = .rightToLeft
        case "radial":
            guard let center = center else { return nil }
            self = .radial(center: center)
        case "inverseradial":
            guard let center = center else { return nil }
            self = .inverseRadial(center: center)
        default:
            return nil
        }
    }
}

// MARK: - CascadePreprocessor

public class CascadePreprocessor: BasePreprocessor {
    // MARK: - Properties

    private let direction: CascadeDirection

    // MARK: - Initialization

    public init(direction: CascadeDirection) {
        self.direction = direction
        super.init()
    }

    public convenience init?(directionString: String, center: CGPoint? = nil) {
        guard let direction = CascadeDirection(string: directionString, center: center) else {
            return nil
        }
        self.init(direction: direction)
    }

    // MARK: - HeroPreprocessor

    public override func process(fromViews: [UIView], toViews: [UIView]) {
        guard let context = context else { return }

        let allViews = fromViews + toViews

        for view in allViews {
            guard let state = context.state(of: view),
                  let cascade = state.cascade else { continue }

            let (delta, cascadeDirection, delayMatchedViews) = cascade

            // Get subviews to cascade
            var subviews = view.subviews

            // Sort subviews based on direction
            subviews = sortViews(subviews, direction: cascadeDirection)

            // Apply cascading delays
            for (index, subview) in subviews.enumerated() {
                if var subviewState = context.state(of: subview) {
                    let additionalDelay = Double(index) * delta
                    let currentDelay = subviewState.delay ?? 0
                    subviewState.delay = currentDelay + additionalDelay
                    context.setState(subviewState, for: subview)
                }
            }
        }
    }

    // MARK: - Helper Methods

    private func sortViews(_ views: [UIView], direction: CascadeDirection) -> [UIView] {
        return views.sorted { (view1, view2) -> Bool in
            let center1 = view1.center
            let center2 = view2.center

            switch direction {
            case .topToBottom:
                return center1.y < center2.y

            case .bottomToTop:
                return center1.y > center2.y

            case .leftToRight:
                return center1.x < center2.x

            case .rightToLeft:
                return center1.x > center2.x

            case .radial(let center):
                let distance1 = distanceSquared(from: center1, to: center)
                let distance2 = distanceSquared(from: center2, to: center)
                return distance1 < distance2

            case .inverseRadial(let center):
                let distance1 = distanceSquared(from: center1, to: center)
                let distance2 = distanceSquared(from: center2, to: center)
                return distance1 > distance2
            }
        }
    }

    private func distanceSquared(from point1: CGPoint, to point2: CGPoint) -> CGFloat {
        let dx = point1.x - point2.x
        let dy = point1.y - point2.y
        return dx * dx + dy * dy
    }
}
